dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/networking/P2PClient.scala

```diff
@@ -0,0 +1,430 @@
+        logger.debug(s"Local: $local")
+
+        //this is what registers a actor to send all byte messages to that is
+        //received from our peer. Since we are using 'self' that means
+        //our bitcoin peer will send all messages to this actor.
+        peerConnection ! Tcp.Register(self)
+        peerConnection ! Tcp.ResumeReading
+
+        currentPeerMsgHandlerRecv =
+          currentPeerMsgHandlerRecv.connect(P2PClient(self, peer))
+        context.become(awaitNetworkRequest(peerConnection, unalignedBytes))
+        unalignedBytes
+
+      case closeCmd @ (Tcp.ConfirmedClosed | Tcp.Closed | Tcp.Aborted |
+          Tcp.PeerClosed) =>
+        logger.info(s"We've been disconnected by $peer command=$closeCmd")
+        //tell our peer message handler we are disconnecting
+        val newPeerMsgRecv = currentPeerMsgHandlerRecv.disconnect()
+
+        currentPeerMsgHandlerRecv = newPeerMsgRecv
+        context.stop(self)
+        unalignedBytes
+
+      case Tcp.Received(byteString: ByteString) =>
+        val byteVec = ByteVector(byteString.toArray)
+        logger.debug(s"Received ${byteVec.length} TCP bytes")
+        logger.trace(s"Received TCP bytes: ${byteVec.toHex}")
+        logger.trace {
+          val post =
+            if (unalignedBytes.isEmpty) "None"
+            else unalignedBytes.toHex
+          s"Unaligned bytes: $post"
+        }
+
+        if (unalignedBytes.isEmpty) {
+          peerConnection ! Tcp.ResumeReading
+        }
+
+        //we need to aggregate our previous 'unalignedBytes' with the new message
+        //we just received from our peer to hopefully be able to parse full messages
+        val bytes: ByteVector = unalignedBytes ++ byteVec
+        logger.trace(s"Bytes for message parsing: ${bytes.toHex}")
+        val (messages, newUnalignedBytes) =
+          P2PClient.parseIndividualMessages(bytes)
+
+        logger.debug {
+          val length = messages.length
+          val suffix = if (length == 0) "" else s": ${messages.mkString(", ")}"
+
+          s"Parsed $length message(s) from bytes$suffix"
+        }
+        logger.debug(s"Unaligned bytes after this: ${newUnalignedBytes.length}")
+        if (newUnalignedBytes.nonEmpty) {
+          logger.trace(s"Unaligned bytes: ${newUnalignedBytes.toHex}")
+        }
+
+        val f: (
+            PeerMessageReceiver,
+            LnMessage[TLV]) => Future[PeerMessageReceiver] = {
+          case (peerMsgRecv: PeerMessageReceiver, m: LnMessage[TLV]) =>
+            logger.trace(s"Processing message=$m")
+            val msg = LnMessageReceived(m, P2PClient(self, peer))
+            peerMsgRecv.handleLnMessageReceived(msg)
+        }
+
+        logger.trace(s"About to process ${messages.length} messages")
+        val newMsgReceiverF =
+          FutureUtil.foldLeftAsync(currentPeerMsgHandlerRecv, messages)(f)(
+            context.dispatcher)
+
+        val newMsgReceiver = Await.result(newMsgReceiverF, timeout)
```

Comment:
That's the kind of legacy I'd like to get rid of.
```scala

def asynchronoslyProcessData(data: ByteString): Future[DataProcessed] = ???

def receive: Receive = {
  case Tcp.Received(data) => asynchronoslyProcessData(data).pipeTo(self)
  case DataProcessed(newMsgReceiver) => currentPeerMsgHandlerRecv = newMsgReceiver
}
```

Reply:
Also it makes sense to make all those receive methods more granular. 
Each step could have it's own `Receive` handler like:

```scala
def sendOffer: Receive = {
  case Send(offer) => 
     sendOffer(offer)
     context become awaitOfferAcceptance
}

def awaitOfferAcceptance: Receive = {
  case OfferAccepted(acc) => 
    processAcc(acc)
    context become nextStep
}

def nextStep: Receive = ???
```

Reply:
That could make the whole cleaner code easier to understand.

Reply:
Also, it look like the the protocol handler is not linear step-by-step one, so it would more appropriate to use Akka's own FSM capabilities: https://doc.akka.io/docs/akka/current/typed/fsm.html

Reply:
Concept ACK on @rorp 's suggestions